src/cargo/core/resolver/mod.rs

```diff
@@ -244,149 +190,240 @@ struct Context {
+    debug!("activating {}", parent.package_id());
 
-    let deps = try!(cx.build_deps(registry, parent, method));
+    let deps = try!(cx.build_deps(registry, &parent, method));
 
-    activate_deps(cx, registry, parent, deps.iter(), 0, &mut |mut cx, registry| {
-        cx.visited.remove(id);
-        finished(cx, registry)
-    })
+    Ok(Some(DepsFrame{
+        parent: parent,
+        remaining_siblings: RcVecIter::new(deps),
+        id: id,
+    }))
 }
 
-/// Activates the dependencies for a package, one by one in turn.
-///
-/// This function will attempt to activate all possible candidates for each
-/// dependency of the package specified by `parent`. The `deps` iterator
-/// provided is an iterator over all dependencies where each element yielded
-/// informs us what the candidates are for the dependency in question.
+#[derive(Clone)]
+struct RcVecIter<Elem> {
+    vec: Rc<Vec<Elem>>,
+    rest: Range<usize>,
+}
+
+impl<Elem> RcVecIter<Elem> {
+    fn new(vec: Vec<Elem>) -> RcVecIter<Elem> {
+        RcVecIter {
+            rest: 0..vec.len(),
+            vec: Rc::new(vec),
+        }
+    }
+}
+impl<Elem> Iterator for RcVecIter<Elem> where Elem: Clone {
+    type Item = (usize, Elem);
+    fn next(&mut self) -> Option<(usize, Elem)> {
+        self.rest.next().and_then(|i| self.vec.get(i).map(|val| (i, val.clone())))
+    }
+}
+
+#[derive(Clone)]
+struct DepsFrame {
+    parent: Rc<Summary>,
+    remaining_siblings: RcVecIter<DepInfo>,
+    id: PackageId,
+}
+
+struct BacktrackFrame {
+    context_backup: Context,
+    deps_backup: Vec<DepsFrame>,
+    remaining_candidates: RcVecIter<Rc<Summary>>,
+    // For building an activation error:
+    parent: Rc<Summary>,
+    cur: usize,
+    dep: Rc<Dependency>,
+    prev_active: Vec<Rc<Summary>>,
+    all_candidates: Vec<Rc<Summary>>,
+}
+
+/// Recursively activates the dependencies for `top`, in depth-first order,
+/// backtracking across possible candidates for each dependency as necessary.
 ///
 /// If all dependencies can be activated and resolved to a version in the
-/// dependency graph the `finished` callback is invoked with the current state
-/// of the world.
-fn activate_deps<'a>(cx: Box<Context>,
-                     registry: &mut Registry,
-                     parent: &Summary,
-                     mut deps: slice::Iter<'a, DepInfo>,
-                     cur: usize,
-                     finished: &mut FnMut(Box<Context>, &mut Registry) -> ResolveResult)
-                     -> ResolveResult {
-    let &(dep, ref candidates, ref features) = match deps.next() {
-        Some(info) => info,
-        None => return finished(cx, registry),
-    };
+/// dependency graph, cx.resolve is returned.
+fn activate_deps_loop(mut cx: Context,
+                      registry: &mut Registry,
+                      top: Rc<Summary>,
+                      top_method: &Method) -> CargoResult<Resolve> {
+    let mut backtrack_stack: Vec<BacktrackFrame> = Vec::new();
```

Comment:
'k, I still disagree and hope the Cargo and Rust communities will reconsider, but done for this change.